sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/ApiServer.java

```diff
@@ -41,17 +42,71 @@
     private Thread serverThread;
 
     public static ApiServer start(
-            SchedulerConfig schedulerConfig, Collection<Object> resources, Runnable startedCallback) {
+            String serviceName,
+            SchedulerConfig schedulerConfig,
+            Collection<Object> resources,
+            Runnable startedCallback) {
         ApiServer apiServer = new ApiServer(schedulerConfig, resources);
         apiServer.start(new AbstractLifeCycle.AbstractLifeCycleListener() {
             @Override
             public void lifeCycleStarted(LifeCycle event) {
+                awaitSchedulerDns(serviceName, schedulerConfig);
                 startedCallback.run();
             }
         });
         return apiServer;
     }
 
+    private static void awaitSchedulerDns(String schedulerName, SchedulerConfig schedulerConfig) {
+        final String schedulerHostname = EndpointUtils.toSchedulerAutoIpHostname(schedulerName, schedulerConfig);
+        final String expectedAddress = schedulerConfig.getSchedulerIP();
+
+        LOGGER.info("Waiting for the scheduler host {} to resolve to {}", schedulerHostname, expectedAddress);
+
+        final Timer resolutionTimer = new Timer("API-resolution-timeout");
+        resolutionTimer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                LOGGER.error("Unable to lookup scheduler host {} within {}ms",
+                        schedulerHostname, schedulerConfig.getApiServerInitTimeout().toMillis());
+                ProcessExit.exit(ProcessExit.API_SERVER_ERROR);
+            }
+        }, schedulerConfig.getApiServerInitTimeout().toMillis());
+
+        while (true) {
+            try {
+                InetAddress[] addresses = InetAddress.getAllByName(schedulerHostname);
+
+                if (addresses.length != 1) {
```

Comment:
Does it make sense to check the multiple addresses for any matching entry?

Reply:
I believe that if there are multiple addresses present, then the DNS is not yet correct.

WDYT, @jeid64?